spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java

```diff
@@ -232,8 +258,8 @@ private void addChannelFromString(Collection<MessageChannel> channels, String ch
 		MessageChannel channel = resolveChannelForName(channelName, message);
 		if (channel != null) {
 			channels.add(channel);
-			if (!mapped && !this.dynamicChannels.contains(channelName)) {
-				this.dynamicChannels.add(channelName);
+			if (!mapped && !(this.dynamicChannels.get(channelName) != null)) {
+				this.dynamicChannels.put(channelName, channel);
```

Comment:
M-m-m. Since we have a cache style now, I think we always should `put` to refresh the old entry.
Otherwise we may just route and instantly be remove by the policy on the next subsequent new route.
No? What am I missing ?